crates/ruff_linter/src/rules/flake8_pyi/rules/bad_generator_return_type.rs

```diff
@@ -176,16 +216,84 @@ pub(crate) fn bad_generator_return_type(
             }
         }
     };
-
-    checker.diagnostics.push(Diagnostic::new(
+    let mut diagnostic = Diagnostic::new(
         GeneratorReturnFromIterMethod {
             better_return_type: better_return_type.to_string(),
             method_name: name.to_string(),
         },
         function_def.identifier(),
-    ));
+    );
+    if let Some(fix) = get_fix(function_def, checker, returns, is_stub, yield_type_info) {
+        diagnostic.set_fix(fix);
+    };
+    checker.diagnostics.push(diagnostic);
 }
 
 fn is_any_or_none(expr: &ast::Expr, semantic: &SemanticModel) -> bool {
     semantic.match_typing_expr(expr, "Any") || matches!(expr, ast::Expr::NoneLiteral(_))
 }
+
+fn get_fix(
+    function_def: &ast::StmtFunctionDef,
+    checker: &Checker,
+    returns: &ast::Expr,
+    is_stub: bool,
+    yield_type_info: Option<YieldTypeInfo>,
+) -> Option<Fix> {
+    let edit = match returns {
+        ast::Expr::Attribute(_) => get_edit(returns),
+        ast::Expr::Subscript(ast::ExprSubscript { value, .. }) => get_edit(value.as_ref()),
+        _ => None,
+    };
+
+    let Some(edit) = edit else {
+        return None;
+    };
+    let mut rest = vec![];
+    if let Some(yield_type_info) = yield_type_info {
+        rest.push(Edit::range_replacement(
+            checker.generator().expr(&yield_type_info.expr),
+            yield_type_info.range,
+        ));
+    }
+
+    // Mark as unsafe if it's a runtime Python file
+    // and the body has more than one statement in it.
+    let applicability = if is_stub || function_def.body.len() == 1 {
+        Applicability::Safe
+    } else {
+        Applicability::Unsafe
+    };
+    Some(Fix::applicable_edits(edit, rest, applicability))
+}
+
+fn get_edit(expr: &ast::Expr) -> Option<Edit> {
+    let ast::Expr::Attribute(ast::ExprAttribute {
+        value, attr, range, ..
+    }) = expr
+    else {
+        return None;
+    };
+
+    let new_return = match attr.as_str() {
+        "Generator" => "Iterator",
+        "AsyncGenerator" => "AsyncIterator",
+        _ => return None,
+    };
+
+    let module = match value.as_ref() {
+        ast::Expr::Name(ast::ExprName { id, .. }) => id.to_owned(),
+        ast::Expr::Attribute(ast::ExprAttribute { attr, value, .. }) => match value.as_ref() {
+            ast::Expr::Name(ast::ExprName { id, .. }) => format!("{id}.{attr}"),
+            _ => return None,
+        },
+        _ => return None,
+    };
+
+    if !["typing", "typing_extensions", "collections.abc"].contains(&module.as_str()) {
+        return None;
+    }
```

Comment:
So, one thing you can explore here: take a look at `get_or_import_symbol` on `Importer`. It lets you define a module-and-member, and then gives you (1) the edit required to import the symbol (if not present), and (2) the bound name of the symbol. It should thus allow you to support cases in which the symbol is not accessed as a fully-qualified name.